New syntax for tuples and lists to remove punning (#475)
ghc-proposals/ghc-proposals: proposals/0000-tuple-syntax.rst

over time.

Proposed Change Specification
-----------------------------

1. Add the following definitions to a new module ``GHC.Tuple.Prim``::

     data Unit = ()
     data Solo a = MkSolo a    -- this is a change from today's `data Solo a = Solo a`

     type Tuple0 = Unit
     type Tuple1 = Solo
     data Tuple2 a b = (a, b)
     data Tuple3 a b c = (a, b, c)
     -- ...
     data Tuple64 ... = (...)

#. Add the following definitions to ``GHC.Classes`` (re-exported from ``GHC.Exts``). These
   replace similar definitions today::

     class CUnit
     instance CUnit

     class a => CSolo a
     instance a => CSolo a

     type CTuple0 = CUnit
     type CTuple1 = CSolo

     class (a, b) => CTuple2 a b
     instance (a, b) => CTuple2 a b

     -- ...

     class (...) => CTuple64 ...
     instance (...) => CTuple64 ...

#. Remove existing tuple definitions from ``GHC.Tuple``.

#. Add the following definitions to ``GHC.Tuple``, which re-exports all of ``GHC.Tuple.Prim``::

     type TupleNKind :: Nat -> Type     -- Nat is from GHC.TypeLits
     type family TupleNKind n = r | r -> n where
       TupleNKind 0 = Type

>        TupleNKind n = Type -> TupleNKind (n-1)   -- this fails the injectivity check, but a little magic will allow this

Reviewer: Do you envision the magic to be specific to `TupleNKind` or all definition that try to use `Nat` inductively?

Author: I think we should do what's necessary to get TupleNKind to compile. If we can do so generally, then all the better.

Right now, GHC has some built-in magic so that, if `n0-1 ~ m` and `n1-1 ~ m`, then GHC will know that `n0 ~ n1`. Note that this magic isn't available to non-built-in functions: it's using the fact that both calls to `-` take the same second argument. But my guess is that this particular existent magic isn't consulted during the injectivity check. So, either we consult it during the injectivity check, or we whip up some custom magic for `TupleNKind`. The former is preferable, but the latter is acceptable. But I don't wish to hold up this proposal on, say, a specification of exactly how to do this.

      for ``GHC.Prim.Tuple``\ *n*\ ``# ty1 ty2 ... tyn-1 tyn``.

   #. An occurrence of ``(# | | ... | | #)`` where there are *n* pipes (for *n* ≧ 1) in type-syntax
      is a synonym for ``GHC.Prim.Sum``\ *n+1*\ ``#``.

   #. An occurrence of ``(# ty1 | ty2 | ... | tyn-1 | tyn #)`` (for *n* ≧ 2) in type-syntax is a
      synonym for ``GHC.Prim.Sum``\ *n*\ ``# ty1 ty2 ... tyn-1 tyn``.

   #. An occurrence of ``()`` in type-syntax, where the type is expected to be of kind ``Constraint``,
      is a synonym for ``GHC.Classes.CUnit``.

   #. An occurrence of ``(ty1, ty2, ..., tyn-1, tyn)`` (for *n* ≧ 2) in type-syntax, where the type is
      expected to be of kind ``Constraint``, is a synonym for ``GHC.Classes.CTuple``\ *n*\ `` ty1 ty2 ... tyn-1 tyn``.

   #. An occurrence of ``(ty1, ty2, ..., tyn-1, tyn)`` (for *n* ≧ 2) in type-syntax, where the first
      ``tyi`` inferred to have kind ``Type`` or ``Constraint`` has kind ``Constraint``, is a synonym
      for ``GHC.Classes.CTuple``\ *n*\ `` ty1 ty2 ... tyn-1 tyn``.

   #. An unapplied occurrence of ``GHC.Types.List`` is pretty-printed as ``[]``.

>    #. An occurrence of ``GHC.Types.List ty`` is pretty-printed as ``[ty]]`.

Reviewer: ```suggestion
   #. An occurrence of ``GHC.Types.List ty`` is pretty-printed as ``[[ty]]``.
```

   #. An occurrence of ``GHC.Tuple.Prim.Tuplen ty1 ty2 ... tyn`` is pretty-printed as ``Tuple [ty1, ty2, ..., tyn]``.

   #. An occurrence of ``GHC.Classes.CTuplen ty1 ty2 ... tyn`` is pretty-printed as ``Constraints [ty1, ty2, ..., tyn]``.

Effect and Interactions
-----------------------
1. With ``-XListTupleTypeSyntax`` (which is on by default), all programs that are accepted today continue
   to be accepted, and with the same meanings. Note that the peculiar dance around type tuples and constraint
   tuples exists today; I have tried to describe the current implementation faithfully, above.

#. With ``-XListTupleTypeSyntax`` (which is on by default), most pretty-printing will happen as it does
   today. The exception is around unsaturated ``CTuplen``, which is not handled above. It is hard to have
   an unsaturated constraint tuple, but possible by the use of a type family that decomposes one. Today's
   GHC prints out e.g. ``ghc-prim-0.6.1:GHC.Classes.(%,%)``. Switching to ``GHC.Classes.CTuple2`` (which is
   actually parseable) seems a positive improvement.

#. With the definitions above, users can avoid puns in their lists and tuples.

   .. _constraints-special-syntax:

> #. Note that the type syntax ``(ty1, ty2, ..., tyn) => ...`` is special syntax. The parser does *not*

Reviewer 2: Oh I meant something simple like
```suggestion
#. Note that the type syntax ``(ty1, ty2, ..., tyn) => ...`` is already special syntax. The parser does *not*
```

type-level namespace or the term-level namespace (but with different meanings).

Furthermore, puns can be hard for newcomers to Haskell. It is not hard to jump
from ::

  single :: [Int]
  single = [5]

to ::

  just :: Maybe Int
  just = Maybe 5

even though the latter is nonsense. With this proposal, users would have the
option of writing ::

  single :: List Int
  single = [5]

Puns also cause trouble when describing what we are talking about: conversations
around these features invariably require people to say "``[]``-the-type" or
"``[]``-the-empty-list" or similar.

This proposal thus allows for a common way to refer to tuples and lists without
punning. It is entirely opt-in, though we expect the new phrasing to become popular
over time.

Proposed Change Specification
-----------------------------

1. Create a new module in ``base`` called ``GHC.Prelude``. This module exports definitions
   peculiar to GHC that are safe to use in ordinary code.
   This module is *not* intended as a replacement to ``Prelude``,
   but instead a complement to it. This module is a counterpart to ``GHC.Exts``, which exports

>    many *unsafe* internals.

Reviewer 3: No, I agree with you. Well, I'd still call this “be imported by default”, if the recommendation is to have it at the top of every file (as opposed to “imported implicitly” for the current `Prelude`). I'm re-reading your sentence, and I'm not sure what I understood yesterday as I read it. So… well… this was noise. Sorry.

   These would probably end up in a new module ``GHC.Tuple.Prim``, due to the dependency from
   the definitions below in ``GHC.Tuple`` on ``GHC.TypeLits``.) ::

     data Unit = ()
     data Solo a = MkSolo a    -- this is a change from today's `data Solo a = Solo a`

     {-# DEPRECATED Solo "The Solo constructor has been renamed to MkSolo to avoid punning." #-}
     pattern Solo :: a -> Solo a
     pattern Solo x = MkSolo x
     {-# COMPLETE Solo #-}

     type Tuple0 = Unit
     type Tuple1 = Solo
     data Tuple2 a b = (a, b)
     data Tuple3 a b c = (a, b, c)
     -- ...
     data Tuple64 ... = (...)

#. Export the following definitions from ``GHC.Exts`` and ``GHC.Prelude``. These
   replace similar definitions today. (Note that ``(...) =>`` is special syntax, and does not
   construct tuples. See more on this point `below <#constraint-special-syntax>`_.)::

     class CUnit
     instance CUnit

     class a => CSolo a
     instance a => CSolo a

     type CTuple0 = CUnit
     type CTuple1 = CSolo

     class (a, b) => CTuple2 a b
     instance (a, b) => CTuple2 a b

     -- ...

     class (...) => CTuple64 ...
     instance (...) => CTuple64 ...

> #. Remove existing tuple definitions from ``GHC.Tuple``.

Author: No, given that, with `-XListTuplePuns`, any type-syntax use of `(...,...)` gets translated to use the new tuples.

     data List a = [] | a : List a

#. Re-export ``List`` from ``GHC.List`` and ``GHC.Prelude``.

#. Introduce a new extension ``-XListTuplePuns``; this extension is part
   of ``-XHaskell98``, ``-XHaskell2010``, and ``-XGHC2021``. It is thus on by default.

#. With ``-XListTuplePuns``:

   1. An occurrence of ``[]`` in type-syntax (as defined in `#378`_) is a synonym
      for ``GHC.List.List``.

   #. An occurrence of ``[ty]`` in type-syntax is a synonym for ``GHC.List.List ty``.

   #. An occurrence of ``()`` in type-syntax, where the type is not expected to be of kind ``Constraint``,
      is a synonym for ``GHC.Tuple.Unit``.

   #. An occurrence of ``(,,...,,)`` where there are *n* commas (for *n* ≧ 1) in type-syntax
      is a synonym for ``GHC.Tuple.Tuple``\ *n+1*.

   #. An occurrence of ``(ty1,ty2,...,tyn-1,tyn)`` (for *n* ≧ 2) in type-syntax, where neither the type
      is expected to be of kind ``Constraint`` and either none of the ``tyi`` are inferred to have kind ``Constraint``
      or there exists a ``tyi`` inferred to kind ``Type`` and none of the ``tyj`` (with j < i) are inferred to have
      kind ``Constraint``, is

>       a synonym for ``GHC.Tuple.Tuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

Author: Yes... but I don't quite know how else to get my point across. Maybe `Tuple<n>`? Would that be an improvement?

>       a synonym for ``GHC.Tuple.Tuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

Reviewer 3: I'd certainly understand `GHC.Tuple.Tuple<n> ty1 ty2 ... tyn-1 tyn` myself. I'm not sure how universal the notation is, though.

>       a synonym for ``GHC.Tuple.Tuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

Author: OK. I agree this is an improvement.

   #. With ``-XUnboxedTuples``, an occurrence of ``(# #)`` in type-syntax is a synonym for ``GHC.Exts.Unit#``.

   #. With ``-XUnboxedTuples``, an occurrence of ``(#,,...,,#)`` where there are *n* commas (for *n* ≧ 1) in type-syntax
      is a synonym for ``GHC.Exts.Tuple``\ *n+1*\ ``#``.

   #. With ``-XUnboxedTuples``, an occurrence of ``(# ty1, ty2, ... , tyn-1, tyn #)`` (for *n* ≧ 2) in type-syntax is a synonym
      for ``GHC.Exts.Tuple``\ *n*\ ``# ty1 ty2 ... tyn-1 tyn``.

   #. With ``-XUnboxedSums``, an occurrence of ``(# | | ... | | #)`` where there are *n* pipes (for *n* ≧ 1) in type-syntax
      is a synonym for ``GHC.Exts.Sum``\ *n+1*\ ``#``.

   #. With ``-XUnboxedSums``, an occurrence of ``(# ty1 | ty2 | ... | tyn-1 | tyn #)`` (for *n* ≧ 2) in type-syntax is a
      synonym for ``GHC.Exts.Sum``\ *n*\ ``# ty1 ty2 ... tyn-1 tyn``.

   #. An occurrence of ``()`` in type-syntax, where the type is expected to be of kind ``Constraint``,
      is a synonym for ``GHC.Tuple.CUnit``.

   #. An occurrence of ``(ty1, ty2, ..., tyn-1, tyn)`` (for *n* ≧ 2) in type-syntax, where the type is

>       expected to be of kind ``Constraint``, is a synonym for ``GHC.Tuple.CTuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

Author: It is what we do today. It *is* scary. But my goal here is not to de-scarify this aspect of GHC -- just to change concrete syntax.

>       expected to be of kind ``Constraint``, is a synonym for ``GHC.Tuple.CTuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

Reviewer 3: I certainly don't intend for you to de-scarify anything :slightly_smiling_face: . Though maybe you want to remind the reader that GHC already does this (because tuples of `Constraint`-s and tuples of `Type`-s are already two different type constructors) and it's merely a syntax change.

>       expected to be of kind ``Constraint``, is a synonym for ``GHC.Tuple.CTuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

Author: So noted.

   does not contain any uses of ``CTuplen``, it will also continue to pretty-print just as today.

   On the other hand, collections of constraints occurring not to the left of a ``=>`` are affected by
   this proposal, for example in ``Dict (Eq a, Show b)`` (which would be written ``Dict (Constraints [Eq a, Show b])``
   under this proposal). Another example is ``(Eq a, (Show a, Read a)) => a -> a``, which would not
   be accepted under ``-XNoListTuplePuns``. Instead, the user should flatten the constraints or
   write ``(Eq a, Constraints [Show a, Read a]) => a -> a``.

#. An instance declaration like ``instance (C a, C b) => C (Tuple [a, b]) where ...`` would be
   rejected because it uses a type family in the instance head. We might choose to relax
   this restriction, by allowing type families in an instance head, as long as they can
   reduce to a ground (i.e. type-family-free) type. This proposal does *not* include such
   a lifting of the restriction, as the workaround is straightforward: just write
   ``instance (C a, C b) => C (Tuple2 a b) where ...``. Still, we may decide to revisit

>    this in the future.

Reviewer 3: Alternatively, you don't define `Tuple2` and such (at least not primitively), and make `Tuple ts` a data family instead. Would this work?

Author: No, because you can't translate what is now `instance Functor ((,) a)`.

-------------------
1. This is one more feature to maintain, but the code would be pretty local.

#. Having multiple ways of naming one thing may offer a boon to *writers* of code
   (they can choose whichever way to name a tuple that they like), but it imposes
   a burden on *readers* of code, who may need to be familiar with all possible
   ways of describing a tuple (and that they are interchangeable). Careful
   documentation of these ideas -- ideally, in the Haddock documentation for the
   names introduced above -- will help to mitigate this problem.

#. A particular class of code readers are beginners, and having multiple different
   ways to say the same thing is particularly challenging for beginners. We should
   thus think carefully about how to present these names to beginners, if
   ``-XNoListTuplePuns`` catches on.

Alternatives
------------
1. Instead of defining ``TupleN`` as a type family (as done here), it could be
   a data family, effectively replacing the ``Tuple2``, ``Tuple3``, ..., definitions.
   This design would seem to be too complicated to be the primitive definition
   of tuples, however, when a very vanilla datatype like ``data Tuple2 a b = (a, b)``

>    suffices.

Reviewer 3: Ha :slightly_smiling_face: .

Still, I think that `Tuple`, is a more reasonable candidate to be made the primitive type than `TupleN`.

#. Instead of introducing new names, we could use more mixfix bits of punctuation,
   such as ``(~ ty1, ty2 ~)`` for normal tuples and ``(% ty1, ty2 %)`` for constraint

>    tuples. This was not as popular in a recent `straw poll <https://github.com/ghc-proposals/ghc-proposals/pull/458#issuecomment-982230541>`_.

Reviewer 3: It's worth noting that MLs do something like this. It's a bit of a weird syntax, though, as `a1 * a2 * … * an` is a single n-ary type constructor (so that, famously, `a * b * c` is not the same type as `a * (b * c)`). No solution is perfect. But I think that it ought to be mentioned in the proposal for completeness.

#. Controlling the ``(# ... | ... | ... #)`` syntax for unboxed sum types with
   ``-XNoListTuplePuns`` is not necessary to avoid punning, but is done only for

>    consistency. We could skip this, but I prefer keeping it as proposed.

Reviewer 3: This is an interesting case. It's not necessary, so if the choice is given, would one prefer using `(# a | b | c #)` or `Sum# [a, b, c]`? Would removing this syntax be considered an anti-feature?

If programmers prefer the former, then it raises an interesting question. It is arguable that unboxed sum are niche enough a feature not to deserve taking up syntactic real estate. However, tuples definitely are. So it would argue that the alternative, above, proposing a special syntax for tuples may have more weight to it than it first seems.

Author: I'd be quite happy with special syntax for tuples. But this seems to be the minority position, if https://github.com/ghc-proposals/ghc-proposals/pull/458#issuecomment-982230541 is to be believed.

> #. There was an objection in the commentary about the name ``GHC.Prelude``. I continue

Reviewer 3: Ha again :slightly_smiling_face:

   but instead a complement to it. This module is a counterpart to ``GHC.Exts``, which exports
   many *unsafe* internals.

   This proposal lists several definitions to be exported from ``GHC.Prelude``, but leaves other
   definitions to be added in separate proposal(s).

#. Export the following definitions from ``GHC.Tuple`` and ``GHC.Prelude``. (Implementation note:
   These would probably end up in a new module ``GHC.Tuple.Prim``, due to the dependency from
   the definitions below in ``GHC.Tuple`` on ``GHC.TypeLits``.) ::

     data Unit = ()
     data Solo a = MkSolo a    -- this is a change from today's `data Solo a = Solo a`

     {-# DEPRECATED Solo "The Solo constructor has been renamed to MkSolo to avoid punning." #-}
     pattern Solo :: a -> Solo a
     pattern Solo x = MkSolo x
     {-# COMPLETE Solo #-}

     getSolo :: Solo a -> a
     getSolo (Solo a) = a   -- as today

     type Tuple0 = Unit
     type Tuple1 = Solo
     data Tuple2 a b = (a, b)
     data Tuple3 a b c = (a, b, c)
     -- ...
     data Tuple64 ... = (...)

#. Export the following definitions from ``GHC.Exts`` and ``GHC.Prelude``. These

>    replace similar definitions (in ``GHC.Classes``) today. (Note that ``(...) =>`` is special syntax, and does not

Reviewer 3: This still sounded a bit mysterious to me. So, for anybody who is not in the know, today (as of Long Ago™), `GHC.Classes` hosts the following class definitions (which are not exported, hence not present in the Haddock). It would be helpful to any future person reading the text if this was explained in the proposal.

```haskell
class ()                                                                                                                                                                                                                                      
class (c1, c2)     => (c1, c2)                                                                                                                                                                                                                
class (c1, c2, c3) => (c1, c2, c3)                                                                                                                                                                                                            
class (c1, c2, c3, c4) => (c1, c2, c3, c4)                                                                                                                                                                                                    
class (c1, c2, c3, c4, c5) => (c1, c2, c3, c4, c5)                                                                                                                                                                                            
class (c1, c2, c3, c4, c5, c6) => (c1, c2, c3, c4, c5, c6)                                                                                                                                                                                    
class (c1, c2, c3, c4, c5, c6, c7) => (c1, c2, c3, c4, c5, c6, c7) 
…
class (c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16,                                                                                                                                                                 
       c17, c18, c19, c20, c21, c22, c23, c24, c25, c26, c27, c28, c29, c30,                                                                                                                                                                  
       c31, c32, c33, c34, c35, c36, c37, c38, c39, c40, c41, c42, c43, c44,                                                                                                                                                                  
       c45, c46, c47, c48, c49, c50, c51, c52, c53, c54, c55, c56, c57, c58,                                                                                                                                                                  
       c59, c60, c61, c62, c63, c64)                                                                                                                                                                                                          
   => (c1, c2, c3, c4, c5, c6, c7, c8, c9, c10, c11, c12, c13, c14, c15, c16,                                                                                                                                                                 
       c17, c18, c19, c20, c21, c22, c23, c24, c25, c26, c27, c28, c29, c30,                                                                                                                                                                  
       c31, c32, c33, c34, c35, c36, c37, c38, c39, c40, c41, c42, c43, c44,                                                                                                                                                                  
       c45, c46, c47, c48, c49, c50, c51, c52, c53, c54, c55, c56, c57, c58,                                                                                                                                                                  
       c59, c60, c61, c62, c63, c64) 
```

Author: I've tried to clarify this.

   construct tuples. See more on this point `below <#constraint-special-syntax>`_.)::

     class CUnit
     instance CUnit

     class a => CSolo a
     instance a => CSolo a

     type CTuple0 = CUnit
     type CTuple1 = CSolo

     class (a, b) => CTuple2 a b
     instance (a, b) => CTuple2 a b

     -- ...

     class (...) => CTuple64 ...
     instance (...) => CTuple64 ...

#. Remove existing tuple definitions from ``GHC.Tuple``.

#. Export the following definitions from ``GHC.Tuple``::

     type TupleNKind :: Nat -> Type     -- Nat is from GHC.TypeLits
     type family TupleNKind n = r | r -> n where
       TupleNKind 0 = Type
       TupleNKind n = Type -> TupleNKind (n-1)   -- this fails the injectivity check, but a little magic will allow this

     type TupleN :: forall (n :: Nat). TupleNKind n
     type family TupleN @n where     -- using syntax from #425
       TupleN @0 = Unit
       TupleN @1 = Solo
       TupleN @2 = Tuple2
       TupleN @3 = Tuple3
       TupleN @4 = Tuple4
       -- ...
       TupleN @64 = Tuple64
       TupleN @n  = TypeError (ShowType n :<>: Text " is too large; the maximum size for a tuple is 64.")

>      type Length :: [a] -> Nat   -- not exported

Author: Indeed! Old habits die hard, even in proposals about removing old habits. Thanks for pointing this out!

   These would likely be exported from ``GHC.Prim`` originally.) ::

     type Unit# :: TYPE (TupleRep [])
     data Unit# = (# #)

     type Solo# :: TYPE rep -> TYPE (TupleRep [rep])
     data Solo# a = MkSolo# a

     type Tuple0# = Unit#
     type Tuple1# = Solo#

     type Tuple2# :: TYPE r1 -> TYPE r2 -> TYPE (TupleRep [r1, r2])
     data Tuple2# a b = (# a, b #)

     -- ...

     type Tuple64# :: TYPE r1 -> ... -> TYPE r64 -> TYPE (TupleRep [r1, ..., r64])
     data Tuple64# ... = (# ... #)

     -- NB: There are no 0-sums or 1-sums in Haskell, today or tomorrow.

     type Sum2# :: TYPE r1 -> TYPE r2 -> TYPE (SumRep [r1, r2])
     data Sum2# a b = (# a | #) | (# | b #)

     type Sum3# :: TYPE r1 -> TYPE r2 -> TYPE r3 -> TYPE (SumRep [r1, r2, r3])
     data Sum3# a b c = (# a | | #) | (# | b | #) | (# | | c #)

     -- ...

     type Sum64# :: TYPE r1 -> ... -> TYPE r64 -> TYPE (SumRep [r1, ..., r64])
     data Sum64# ... = ...

#. Change ``GHC.Types`` to have the following definition::

     data List a = [] | a : List a

#. Re-export ``List`` from ``GHC.List`` and ``GHC.Prelude``.

#. Introduce a new extension ``-XListTuplePuns``; this extension is part

>    of ``-XHaskell98``, ``-XHaskell2010``, and ``-XGHC2021``. It is thus on by default.

Author: This is stated below, but perhaps too late. I've added a comment right at the top.

      for ``GHC.List.List``.

   #. An occurrence of ``[ty]`` in type-syntax is a synonym for ``GHC.List.List ty``.

   #. An occurrence of ``()`` in type-syntax, where the type is not expected to be of kind ``Constraint``,
      is a synonym for ``GHC.Tuple.Unit``.

   #. An occurrence of ``(,,...,,)`` where there are *n* commas (for *n* ≧ 1) in type-syntax
      is a synonym for ``GHC.Tuple.Tuple``\ *n+1*.

   #. An occurrence of ``(ty1,ty2,...,tyn-1,tyn)`` (for *n* ≧ 2) in type-syntax, where neither the type
      is expected to be of kind ``Constraint`` and either none of the ``tyi`` are inferred to have kind ``Constraint``
      or there exists a ``tyi`` inferred to kind ``Type`` and none of the ``tyj`` (with j < i) are inferred to have
      kind ``Constraint``, is
      a synonym for ``GHC.Tuple.Tuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

   #. With ``-XUnboxedTuples``, an occurrence of ``(# #)`` in type-syntax is a synonym for ``GHC.Exts.Unit#``.

   #. With ``-XUnboxedTuples``, an occurrence of ``(#,,...,,#)`` where there are *n* commas (for *n* ≧ 1) in type-syntax
      is a synonym for ``GHC.Exts.Tuple``\ *n+1*\ ``#``.

> 

Author: Quite right. Thanks!

      is a synonym for ``GHC.Exts.Sum``\ *n+1*\ ``#``.

   #. With ``-XUnboxedSums``, an occurrence of ``(# ty1 | ty2 | ... | tyn-1 | tyn #)`` (for *n* ≧ 2) in type-syntax is a
      synonym for ``GHC.Exts.Sum``\ *n*\ ``# ty1 ty2 ... tyn-1 tyn``.

   #. An occurrence of ``()`` in type-syntax, where the type is expected to be of kind ``Constraint``,
      is a synonym for ``GHC.Tuple.CUnit``.

   #. An occurrence of ``(ty1, ty2, ..., tyn-1, tyn)`` (for *n* ≧ 2) in type-syntax, where the type is
      expected to be of kind ``Constraint``, is a synonym for ``GHC.Tuple.CTuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

   #. An occurrence of ``(ty1, ty2, ..., tyn-1, tyn)`` (for *n* ≧ 2) in type-syntax, where the first
      ``tyi`` inferred to have kind ``Type`` or ``Constraint`` has kind ``Constraint``, is a synonym
      for ``GHC.Tuple.CTuple``\ *n* ``ty1 ty2 ... tyn-1 tyn``.

   #. An unapplied occurrence of ``GHC.List.List`` is pretty-printed as ``[]``.

   #. An occurrence of ``GHC.List.List ty`` is pretty-printed as ``[ty]``.

   #. An occurrence of ``GHC.Tuple.Unit`` is pretty-printed as ``()``.

>    #. An occurrence of ``GHC.Tuple.Tuplen ty1 ty2 ... tyn`` is pretty-printed as ``(ty1, ty2, ..., tyn)``.

Author: It's actually correct as it was, where the `n` was a variable. I think this is clearer in the new version.

   peculiar to GHC that are safe to use in ordinary code.
   This module is *not* intended as a replacement to ``Prelude``,
   but instead a complement to it. This module is a counterpart to ``GHC.Exts``, which exports
   many *unsafe* internals.

   This proposal lists several definitions to be exported from ``GHC.Prelude``, but leaves other
   definitions to be added in separate proposal(s).

#. Export the following definitions from ``GHC.Tuple`` and ``GHC.Prelude``. (Implementation note:
   These would probably end up in a new module ``GHC.Tuple.Prim``, due to the dependency from
   the definitions below in ``GHC.Tuple`` on ``GHC.TypeLits``.) ::

     data Unit = ()
     data Solo a = MkSolo a    -- this is a change from today's `data Solo a = Solo a`

     {-# DEPRECATED Solo "The Solo constructor has been renamed to MkSolo to avoid punning." #-}
     pattern Solo :: a -> Solo a
     pattern Solo x = MkSolo x
     {-# COMPLETE Solo #-}

     getSolo :: Solo a -> a
     getSolo (Solo a) = a   -- as today

     type Tuple0 = Unit
     type Tuple1 = Solo
     data Tuple2 a b = (a, b)
     data Tuple3 a b c = (a, b, c)
     -- ...
     data Tuple64 ... = (...)

> #. Export the following definitions from ``GHC.Exts`` and ``GHC.Prelude``. These
>    the existing tuple definitions (in ``GHC.Classes``) today. (Note that ``(...) =>`` is special syntax, and does not

Reviewer 4: ```suggestion
#. Export the following definitions from ``GHC.Exts`` and ``GHC.Prelude``. These replace
   the existing tuple definitions (in ``GHC.Classes``) today. (Note that ``(...) =>`` is special syntax, and does not
```

   #. An occurrence of ``()`` in type-syntax, where the type is not expected to be of kind ``Constraint``,
      is a synonym for ``GHC.Tuple.Unit``.

   #. An occurrence of ``(,,...,,)`` where there are *n* commas (for *n* ≧ 1) in type-syntax
      is a synonym for ``GHC.Tuple.Tuple<n+1>``.

   #. An occurrence of ``(ty1,ty2,...,ty<n-1>,ty<n>)`` (for *n* ≧ 2) in type-syntax, where neither the type
      is expected to be of kind ``Constraint`` and either none of the ``ty<i>`` are inferred to have kind ``Constraint``
      or there exists a ``ty<i>`` inferred to kind ``Type`` and none of the ``ty<j>`` (with *j* < i) are inferred to have
      kind ``Constraint``, is
      a synonym for ``GHC.Tuple.Tuple<n>`` ``ty1 ty2 ... ty<n-1> ty<n>``. (This rule retains today's behavior.)

   #. With ``-XUnboxedTuples``, an occurrence of ``(# #)`` in type-syntax is a synonym for ``GHC.Exts.Unit#``.

   #. With ``-XUnboxedTuples``, an occurrence of ``(# ty #)`` in type-syntax is a synonym for ``GHC.Exts.Solo# ty``.

   #. With ``-XUnboxedTuples``, an occurrence of ``(#,,...,,#)`` where there are *n* commas (for *n* ≧ 1) in type-syntax

>       is a synonym for ``GHC.Exts.Tuple<n+1>#.

Reviewer 4: ```suggestion
      is a synonym for ``GHC.Exts.Tuple<n+1>#``.
```

      never types or type constructors. (Note that ``(...) =>`` is special syntax, not an occurrence of any of the types
      listed above. See `below <#constraints-special-syntax>`_.)

   #. A use of ``(# ... | ... | ... #)``, where each of the ``...`` is filled in, (among other arities) is now disallowed.

   #. An occurrence of ``GHC.Tuple.Tuple<n> ty1 ty2 ... ty<n>`` is pretty-printed as ``Tuple [ty1, ty2, ..., ty<n>]``.

   #. An occurrence of ``GHC.Tuple.CTuple<n> ty1 ty2 ... ty<n>`` is pretty-printed as ``Constraints [ty1, ty2, ..., ty<n>]``.

#. Three releases after this proposal is implemented, remove the ``Solo`` pattern synonym from ``GHC.Tuple``.

Effect and Interactions
-----------------------
1. With ``-XListTuplePuns`` (which is on by default), all programs that are accepted today continue
   to be accepted, and with the same meanings. Note that the peculiar dance around type tuples and constraint
   tuples exists today; I have tried to describe the current implementation faithfully, above.

#. With ``-XListTuplePuns`` (which is on by default), most pretty-printing will happen as it does

>    today. The exception is around unsaturated ``CTuplen``, which is not handled above. It is hard to have

Reviewer 4: ```suggestion
   today. The exception is around unsaturated ``CTuple<n>``, which is not handled above. It is hard to have
```

  (Actually, we would probably not expect these at all, but the current design allows us to be forgiving
  during refactoring.)

  In ``Tuple``, we see ``Tuple a = a``, which looks like a universally applicable equation. It is
  not. Because ``a :: Type`` (resp. ``a :: Constraint``) we learn the invisible argument to ``Tuple``
  (resp. ``Constraints``) must be ``1``, and thus this equation fires only when the argument to ``Tuple``
  is not a tuple of types (resp. constraints).

  A beautiful, unexpected consequence of this design is that it aids migration. Now, with or without
  ``-XListTuplePuns``, a user can write e.g. ``Tuple (Int, Bool, Double)``, and this will be the types
  of e.g. ``(1, True, 3.14)``. With ``-XListTuplePuns``, the argument to ``Tuple`` will be the tuple
  type, of kind ``Type``. That is, the type will really be understood as ``Tuple (Tuple3 Int Bool Double)``.
  The "``1``" equation fires, reducing to ``Tuple3 Int Bool Double``. On the other hand, with
  ``-XNoListTuplePuns``, the user's type is understood as ``Tuple @3 (Int, Bool, Double)``, and the "``3``"
  equation fires, reducing to ``Tuple3 Int Bool Double`` -- the same answer! And so, all users

>   can write ``Tuple`` before their tuples and not get hurt.

Reviewer 4: This part needs to be indented by one additional space, currently it is not displayed correctly.

       TupleArgKind# other = TypeError (ShowType (Length other) :<>: Text " is too large; the maximum size of a tuple is 64.")

     type Tuple# :: forall (reps :: List RuntimeRep). TupleArgKind# reps -> TYPE (TupleRep reps)
     type family Tuple# ts where
       Tuple# () = Unit#
       Tuple# (a :: TYPE r) = TypeError (Text "Tuple# does not work for 1-tuples; use Solo#.")  -- see below
       Tuple# (a, b) = Tuple2# a b
       Tuple# (a, b, c) = Tuple3# a b c
       -- ...
       Tuple# (a, b, ..., bk, bl) = Tuple64# a b ... bk bl
       Tuple# @reps _ = TypeError (ShowType (Length reps) :<>: Text " is too large; the maximum size of a tuple is 64.")

     type Sum# :: forall (reps :: List RuntimeRep). TupleArgKind# reps -> TYPE (SumRep reps)
     type family Sum# ts where
       Sum# () = TypeError (Text "GHC does not support empty unboxed sums. Consider Data.Void.Void instead.")
       Sum# (a :: TYPE r) = TypeError (Text "GHC does not support unary unboxed sums. Consider Data.Tuple.Solo# instead.")
       Sum# (a, b) = Sum2# a b
       Sum# (a, b, c) = Sum3# a b c
       -- ...
       Sum# (a, b, ..., bk, bl) = Sum64# a b ... bk bl

>        Sum# @reps _ = TypeError (ShowType (Length reps) :<>: Text " is too large; the maximum size of a sum is 64.")

Reviewer 5: Today, unboxed sums are not limited to 64 alternatives.

```
ghci> :t (# () | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | #)
(# () | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | | #)
  :: (# () | b | c | d | e | f | g | h | i | j | k | l | m | n | o |
        p | q | r | s | t | u | v | w | x | y | z | t26 | t27 | t28 | t29 |
        t30 | t31 | t32 | t33 | t34 | t35 | t36 | t37 | t38 | t39 | t40 |
        t41 | t42 | t43 | t44 | t45 | t46 | t47 | t48 | t49 | t50 | t51 |
        t52 | t53 | t54 | t55 | t56 | t57 | t58 | t59 | t60 | t61 | t62 |
        t63 | t64 | t65 | t66 | t67 | t68 | t69 | t70 | t71 | t72 | t73 |
        t74 | t75 | t76 | t77 | t78 | t79 | t80 | t81 | t82 | t83 | t84 |
        t85 | t86 | t87 | t88 | t89 | t90 | t91 | t92 | t93 | t94 | t95 |
        t96 | t97 | t98 | t99 | t100 | t101 | t102 | t103 | t104 | t105 |
        t106 | t107 | t108 | t109 | t110 | t111 | t112 | t113 | t114 | t115
        | t116 | t117 | t118 | t119 | t120 | t121 | t122 | t123 | t124 |
        t125 | t126 | t127 | t128 #)
```

That's not a major problem for this `Sum#` type family because it can't be given as input a tuple wider than 64 entries anyway. But it poses other problems:

* The `Sum<n>#` family of type constructors exported from `Data.Sum.Experimental` is not complete.
* Pretty-printing of an unboxed sum type with more than 64 alternatives cannot use the `Sum#` type family with `-XNoListTuplePuns`.
* Because the `(# | | | #)` syntax is disallowed, it is not obvious how to write a "large" unboxed sum type or term when `-XNoListTuplePuns` is active.
